slang.h

```diff
@@ -984,6 +1048,15 @@ extern "C"
         SlangDiagnosticCallback callback,
         void const*             userData);
 
+    SLANG_API void spSetWriter(
+        SlangCompileRequest*    request,
+        SlangWriterChannel      channel, 
+        ISlangWriter*           writer);
```

Comment:
Well as it was, code within slang would write stuff out to diagnostic, stderr, stdout. 
I think that does make sense - it's somewhat regrettable that it is doing this directly (and this change resolves that), but I think if I want to dump IR - it not doing it to diagnostics doesn't seem wrong?
I think this is for multiple reasons - but one way of talking about this would be say another way of potentially implementing the functionality. For example you could have another parameter to the diagnostic callback that says what it is. This would make the 'diagnositic' call back, be the hub of multiplexing the output. This doesn't work in a few ways
* We couldn't have null terminated strings - because that doesn't make sense for binary data
* If different output paths can be different kind of strings (again like they are now), there is no mechanism to query that, or write things in ways appropriate for that target (implying the callback should be an interface)
Having multiple channels and an interface for streams is arguably better from a flexibility/extendability point of view than say a single interface that i stateful and deplexes.

Reply:
There's really two classes of output here:

1. The "primary" compiler output, which is the generated code/assembly/whataever.

2. The "secondary" output, which is diagnostics, dumping, etc.

I think that in the long run all this I/O redirection can be scrapped if we just have a simple policy that:

1. When testing, primary output is all pulled through the established API entry points for retrieving generated code, rather than by running the compiler in "command-line mode" and interception stdout.

2. All secondary output should go to the diagnostic sink/callback, at which point there is already a working API for pulling it and there is again no need for interception.

In the case of (2), this means that it is a Good Thing to migrate the Slang code away from directly using stderr/stdout. In the case of (1) the code that had previously been directly using stdout for writing output should eventually be reverted to doing so, without an interception layer, since that code is specific to command-line `slangc` (it could eventually be moved into `slangc` proper, if we think it is important for `slang.dll` to avoid console I/O completely).